react/components/NewsletterContext.tsx

```diff
@@ -0,0 +1,163 @@
+import { ApolloError } from 'apollo-client'
+import React, {
+  createContext,
+  useReducer,
+  useContext,
+  PropsWithChildren,
+  useEffect,
+} from 'react'
+import { MutationFunction, useMutation } from 'react-apollo'
+
+import subscribeNewsletterMutation from '../graphql/subscribeNewsletter.gql'
+
+interface UpdateEmailAction {
+  type: 'UPDATE_EMAIL'
+  value: string
+}
+interface UpdateNameAction {
+  type: 'UPDATE_NAME'
+  value: string
+}
+
+interface SetInvalidEmailAction {
+  type: 'SET_INVALID_EMAIL'
+  value: boolean
+}
+
+interface SetInvalidNameAction {
+  type: 'SET_INVALID_NAME'
+  value: boolean
+}
+
+interface SetMutationValues {
+  type: 'SET_MUTATION_VALUES'
+  value: {
+    mutationResult: boolean | undefined
+    mutationLoading: boolean
+    mutationError: ApolloError | undefined
+  }
+}
+
+interface State {
+  name: string
+  email: string
+  invalidEmail: boolean
+  invalidName: boolean
+  mutationResult: boolean | undefined
+  mutationLoading: boolean
+  mutationError: ApolloError | undefined
+  subscribeMutation: MutationFunction<boolean, { email: string; name?: string }>
+}
```

Comment:
```suggestion
type SubmissionState = SubmissionError | SubmissionSuccess | SubmissionLoading

interface SubmissionError {
  error: ApolloError
  data: undefined
  loading: false
}

interface SubmissionSuccess {
  error: undefined
  data: true
  loading: false
}

interface SubmissionLoading {
  error: undefined
  data: undefined
  loading: true
}

interface State {
  name: string
  email: string
  invalidEmail: boolean
  invalidName: boolean
  submission: SubmissionState
  subscribe: MutationFunction<boolean, { email: string; name?: string }>
}
```

Reply:
This looks really good! But I could not make it work in with this exact implementation, as It messes up the type inferring from TypeScript and React types for the `useReducer` function.

But I did simplify the context 😄